urllib3/filepost.py

```diff
@@ -56,39 +53,165 @@ def iter_fields(fields):
     return ((k, v) for k, v in fields)
 
 
-def encode_multipart_formdata(fields, boundary=None):
-    """
-    Encode a dictionary of ``fields`` using the multipart/form-data MIME format.
+def encode(string):
+    if isinstance(string, six.text_type):
+        return string.encode('utf-8')
+    else:
+        return string
 
-    :param fields:
-        Dictionary of fields or list of (key, :class:`~urllib3.fields.RequestField`).
 
-    :param boundary:
-        If not specified, then a random boundary will be generated using
-        :func:`mimetools.choose_boundary`.
+def file_size(fp):
+    pos = fp.tell()
+    fp.seek(0, 2)
+    size = fp.tell()
+    fp.seek(pos)
+    return size-pos
+
+
+class IterStreamer(object):
+    """
+    File-like streaming iterator.
     """
-    body = BytesIO()
-    if boundary is None:
-        boundary = choose_boundary()
+    def __init__(self, generator):
+        self.generator = generator
+        self.iterator = iter(generator)
+        self.leftover = b''
 
-    for field in iter_field_objects(fields):
-        body.write(b('--%s\r\n' % (boundary)))
+    def __len__(self):
+        return self.generator.__len__()
 
-        writer(body).write(field.render_headers())
-        data = field.data
+    def __iter__(self):
+        return self.iterator
 
-        if isinstance(data, int):
-            data = str(data)  # Backwards compatibility
+    def next(self):
+        return six.next(self.iterator)
 
-        if isinstance(data, six.text_type):
-            writer(body).write(data)
-        else:
-            body.write(data)
+    def __next__(self):
+        return self.next()
+
+    def read(self, size=None):
+        data = self.leftover
+        count = len(self.leftover)
+        try:
+            while count < size:
+                chunk = self.next()
+                data += chunk
+                count += len(chunk)
+        except StopIteration:
+            pass
 
-        body.write(b'\r\n')
+        if count > size:
+            self.leftover = data[size:]
 
-    body.write(b('--%s--\r\n' % (boundary)))
+        return data[:size]
```

Comment:
It might be nice if this had `str`/`unicode` methods so that it could behave like a string in circumstances where people want to use it like one. Just a thought.